File: app/src/main/java/droiddevelopers254/droidconke/views/activities/AuthenticateUserActivity.java
package droiddevelopers254.droidconke.views.activities;

import android.app.ProgressDialog;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import com.firebase.ui.auth.AuthUI;
import com.firebase.ui.auth.ErrorCodes;
import com.firebase.ui.auth.IdpResponse;
import com.google.android.gms.common.SignInButton;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.ontbee.legacyforks.cn.pedant.SweetAlert.SweetAlertDialog;

import java.util.Collections;

import butterknife.BindView;
import droiddevelopers254.droidconke.HomeActivity;
import droiddevelopers254.droidconke.R;
import droiddevelopers254.droidconke.models.UserModel;
import droiddevelopers254.droidconke.viewmodels.AuthenticateUserViewModel;

public class AuthenticateUserActivity extends AppCompatActivity {

    private static final int RC_SIGN_IN = 123;
    FirebaseAuth auth;
    SignInButton googleSignInBtn;
    private SweetAlertDialog pDialog;
    @BindView(R.id.coorAuthUser)
    CoordinatorLayout snackBarView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //set layout tu fullcreen
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        //transparent status bar
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            getWindow().getDecorView().setSystemUiVisibility(
                    View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                            | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN);
            getWindow().setStatusBarColor(ContextCompat.getColor(this, R.color.mdtp_transparent_black));
        }

        setContentView(R.layout.activity_authenticate_user);

        // Progress dialog
        pDialog = new SweetAlertDialog(AuthenticateUserActivity.this,SweetAlertDialog.PROGRESS_TYPE);
        pDialog.getProgressHelper().setBarColor(Color.parseColor("#863B96"));
        pDialog.setCancelable(false);

        googleSignInBtn=findViewById(R.id.googleSignInBtn);

        //check whether the user is signed in first
         auth = FirebaseAuth.getInstance();
        if (auth.getCurrentUser() != null) {
            // already signed in
            navigateToHome();
        } else {
            // not signed in
           showUI();
        }
    }
    private void showUI() {
        googleSignInBtn.setOnClickListener(view -> {
            pDialog.setTitleText("Signing in");
            showDialog();
            signInUser();
        });
    }

    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        // RC_SIGN_IN is the request code you passed into startActivityForResult(...) when starting the sign in flow.
        if (requestCode == RC_SIGN_IN) {
            hideDialog();
            IdpResponse response = IdpResponse.fromResultIntent(data);

            // Successfully signed in
            if (resultCode == RESULT_OK) {
                //save the user now to db
                navigateToHome();

            } else {
                // Sign in failed
                if (response == null) {
                    // UserModel pressed back button
                   Toast.makeText(getApplicationContext(),"You pressed back button before log in",Toast.LENGTH_SHORT).show();
                    return;
                }

                if (response.getErrorCode() == ErrorCodes.NO_NETWORK) {
                   Toast.makeText(getApplicationContext(),"Network Error",Toast.LENGTH_SHORT).show();
                    return;
                }

                if (response.getErrorCode() == ErrorCodes.UNKNOWN_ERROR) {
                   Toast.makeText(getApplicationContext(),"Please try again",Toast.LENGTH_SHORT).show();

                }
            }

        }
    }
    //function to log in
    private void signInUser(){

        startActivityForResult(
                AuthUI.getInstance()
                        .createSignInIntentBuilder()
                        .setIsSmartLockEnabled(false)
                        .setAvailableProviders(
                                Collections.singletonList(
                                        new AuthUI.IdpConfig.GoogleBuilder().build()
                                ))
                        .build(),
                RC_SIGN_IN);
    }

    private void showSnackbar(String message){
        Snackbar.make(snackBarView,message,Snackbar.LENGTH_SHORT).show();

    }
    private void navigateToHome() {
        Intent intent = new Intent(AuthenticateUserActivity.this,HomeActivity.class);
        startActivity(intent);
        finish();

    }

    private void showDialog() {
        if (!pDialog.isShowing())
            pDialog.show();
    }

    private void hideDialog() {
        if (pDialog.isShowing())
            pDialog.dismiss();

    } 

}

File: app/src/main/java/droiddevelopers254/droidconke/views/activities/AuthenticateUserActivity.kt
package droiddevelopers254.droidconke.views.activities

import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.Window
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.ontbee.legacyforks.cn.pedant.SweetAlert.SweetAlertDialog
import droiddevelopers254.droidconke.HomeActivity
import droiddevelopers254.droidconke.R
import kotlinx.android.synthetic.main.content_authenticate_user.*
import org.jetbrains.anko.toast

class AuthenticateUserActivity : AppCompatActivity() {
    lateinit var auth: FirebaseAuth
    private var pDialog: SweetAlertDialog? = null
    private lateinit var googleSignInClient: GoogleSignInClient

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.AppTheme_NoActionBar)
        super.onCreate(savedInstanceState)

        //set layout tu fullscreen
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN)
        //transparent status bar
        when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP -> {
                window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                window.statusBarColor = ContextCompat.getColor(this, R.color.mdtp_transparent_black)
            }
        }

        setContentView(R.layout.activity_authenticate_user)

        // Configure Google Sign In
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build()
        googleSignInClient = GoogleSignIn.getClient(this, gso)


        // Progress dialog
        pDialog = SweetAlertDialog(this@AuthenticateUserActivity, SweetAlertDialog.PROGRESS_TYPE)
        pDialog?.progressHelper?.barColor = Color.parseColor("#863B96")
        pDialog?.setCancelable(false)

        //check whether the user is signed in first
        auth = FirebaseAuth.getInstance()
        if (auth.currentUser != null) {
            // already signed in
            navigateToHome()
        } else {
            // not signed in
            showUI()
        }
    }

    private fun showUI() {
        googleSignInBtn.setOnClickListener {
            pDialog?.titleText = "Signing in"
            showDialog()
            signInUser()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            RC_SIGN_IN -> {
                val task = GoogleSignIn.getSignedInAccountFromIntent(data)
                try {
                    // Google Sign In was successful, authenticate with Firebase
                    val account = task.getResult(ApiException::class.java)
                    account?.let { firebaseAuthWithGoogle(it) }
                } catch (e: ApiException) {
                    // Google Sign In failed
                    hideDialog()
                    toast("Google sign in failed")
                }
            }
        }
    }

    private fun firebaseAuthWithGoogle(account: GoogleSignInAccount) {
        val credential = GoogleAuthProvider.getCredential(account.idToken, null)
        auth.signInWithCredential(credential)
                .addOnCompleteListener(this) {
                    when {
                        it.isSuccessful -> {
                            // Sign in success, update UI with the signed-in user's information
                            val user = auth.currentUser
                            navigateToHome()
                        }
                        else -> // If sign in fails, display a message to the user.
                            toast("Authentication Failed.")
                    }
                    hideDialog()
                }
    }

    //function to log in
    private fun signInUser() {
        val signInIntent = googleSignInClient.signInIntent
        startActivityForResult(signInIntent, RC_SIGN_IN)
    }

    private fun navigateToHome() {
        val intent = Intent(this@AuthenticateUserActivity, HomeActivity::class.java)
        startActivity(intent)
        finish()

    }

    private fun showDialog() {
        when {
            !pDialog!!.isShowing -> pDialog?.show()
        }
    }

    private fun hideDialog() {
        when {
            pDialog!!.isShowing -> pDialog?.dismiss()
        }

    }

    companion object {
        private const val RC_SIGN_IN = 9001
    }
}
